csrc/sycl/kernels.cpp

```diff
@@ -0,0 +1,5528 @@
+  data[threadid] = output;
+}
+
+template<typename T, int BLOCK_SIZE, int NUM_MAX>
+void kCompressMax(T * __restrict__ const A, T* out, unsigned char* out_idx, const int n,
+                  const sycl::nd_item<3> &item_ct1, int *smem_max_indices,
+                  float *smem_max_values)
+{
+  
+  
+  //typename WarpReduce::TempStorage temp_storage;
+  //typedef cub::BlockLoad<T, BLOCK_SIZE/8 , 8, cub::BLOCK_LOAD_WARP_TRANSPOSE> LoadT;
+  //typename LoadT::TempStorage loadt;
+
+  const int warp_idx = item_ct1.get_local_id(2)/32;
+  const int valid_items = n - (item_ct1.get_group(2)*BLOCK_SIZE) > BLOCK_SIZE ? BLOCK_SIZE : n - (item_ct1.get_group(2)*BLOCK_SIZE);
+
+  //  BLOCK_SIZE/32 == number of warps
+  T values[8];
+  T max1 = -64000.0f;
+  T max2 = -64000.0f;
+  int max_idx1 = -1;
+  int max_idx2 = -1;
+  int sign1 = -1;
+  int sign2 = -1;
+
+  sycl::buffer<T, 1>  buff_indices(smem_max_indices, sycl::range<1>(8*BLOCK_SIZE/32));
+  sycl::buffer<T, 1>  buff_values(smem_max_values, sycl::range<1>(8*BLOCK_SIZE/32));
+  sycl::buffer<T, 1>  buff_A(A,sycl::range<1>(8*BLOCK_SIZE/32));
+  
+  dpct::get_in_order_queue().submit([&](sycl::handler &h) {
+ 
+ 
+      using group_load = dpct::group::workgroup_load<BLOCK_SIZE/8,BLOCK_LOAD_DIRECT>;
+      size_t temp_storage_size = group_load::get_local_memory_size(8*BLOCK_SIZE/32);
+      sycl::local_accessor<uint8_t, 1> tacc(
+        temp_storage_size, h);
+      sycl::accessor dacc(buff_A[(item_ct1.get_local_id(2)*BLOCK_SIZE)], h, sycl::read_write);
+      
+      // 1. load 8 values per thread
+      // 2. compute 2-max in registers (64 max per warp)
+      // 3. do warp reduction + broadcast back
+      // 4. Up-shift maxed value, write index into shared memory, replace with 2nd largest
+      // 5. Repeat (3) 8 times for top 8 values in 256
+      // 6. store with byte index
+      h.parallel_for(sycl::nd_range<3>(sycl::range<3>(1, 1, 128), sycl::range<3>(1, 1, 128)),
+        [=](sycl::nd_item<3> item) {
+          auto *d = dacc.get_multi_ptr<sycl::access::decorated::yes>().get();
+          auto *tmp = tacc.get_multi_ptr<sycl::access::decorated::yes>().get();
+          group_load(tmp).load(item_ct1,item_ct1.get_local_linear_id(), d, values);
+        
+        });
+      
+  });
+   
+  #pragma unroll 8
+  for(int i = 0; i < 8; i++)
+  {
+    T absval = fabsf(values[i]);
+    if(absval > max1)
+    {
+      max1 = values[i];
+      sign1 = signbit(values[i]);
+      max_idx1 = 8*item_ct1.get_local_id(2) + i;
+    }
+    else if(absval > max2)
+    {
+      max2 = values[i];
+      sign2 = signbit(values[i]);
+      max_idx2 = 8*item_ct1.get_local_id(2) + i;
+    }
+  }
+  
+  float warp_max;
+  sycl::host_accessor hacc_values{buff_values};
+  sycl::host_accessor hacc_indices{buff_indices};
+  for(int i = 0; i < 8; i++)
+  {
+    // 3. do warp reduction + broadcast back
+    
+    output = sycl::reduce_over_group(item_ct1.get_sub_group(), max1, sycl::maximum<>());
```

Comment:
Same here re portability, if you ported this directly from cuda code you can only expect it be functionally identical on devices of subgroup size 32.

Reply:
I think we can just remove `kCompressMax` as it does not appear to be used at all. There's a few other areas where there might be some dead code, at least on the CUDA side where this is sourced from.

Reply:
Yes kcompressmax (+ some other cuda code) is pretty much unused. I will clean up where-ever dead code resides.